sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -1,103 +1,247 @@
 // Copyright 2021-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use alloc::{boxed::Box, string::String, vec::Vec};
-use core::{ops::RangeInclusive, str::FromStr};
-
-use packable::{bounded::BoundedU16, prefix::BoxedSlicePrefix};
+use alloc::{
+    borrow::ToOwned,
+    collections::BTreeMap,
+    format,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::ops::RangeInclusive;
+
+use packable::{
+    bounded::{BoundedU16, BoundedU8},
+    error::{UnpackError, UnpackErrorExt},
+    packer::Packer,
+    prefix::{BTreeMapPrefix, BoxedSlicePrefix},
+    unpacker::{CounterUnpacker, Unpacker},
+    Packable, PackableExt,
+};
 
 use crate::types::block::{output::StorageScore, protocol::WorkScore, Error};
 
-pub(crate) type MetadataFeatureLength =
-    BoundedU16<{ *MetadataFeature::LENGTH_RANGE.start() }, { *MetadataFeature::LENGTH_RANGE.end() }>;
+pub(crate) type MetadataFeatureEntryCount = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>;
+pub(crate) type MetadataFeatureValueLength = BoundedU16<0, { u16::MAX }>;
+
+type MetadataBTreeMapPrefix = BTreeMapPrefix<
+    BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+    BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+    MetadataFeatureEntryCount,
+>;
+
+type MetadataBTreeMap =
+    BTreeMap<BoxedSlicePrefix<u8, MetadataFeatureKeyLength>, BoxedSlicePrefix<u8, MetadataFeatureValueLength>>;
 
 /// Defines metadata, arbitrary binary data, that will be stored in the output.
-#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, packable::Packable)]
-#[packable(unpack_error = Error, with = |err| Error::InvalidMetadataFeatureLength(err.into_prefix_err().into()))]
-pub struct MetadataFeature(
-    // Binary data.
-    pub(crate) BoxedSlicePrefix<u8, MetadataFeatureLength>,
-);
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct MetadataFeature(MetadataBTreeMapPrefix);
+
+fn verify_packable<const VERIFY: bool>(feature: &MetadataFeature) -> Result<(), Error> {
+    verify_keys_packable::<VERIFY>(feature)?;
+    verify_byte_length_packable::<VERIFY>(feature.packed_len())?;
```

Comment:
```suggestion
    // Add feature type byte
    verify_byte_length_packable::<VERIFY>(feature.packed_len() + 1)?;
```

Reply:
I read the TIP again and the range is only for the map inside without the type, so it's not needed (also asked Philipp and he confirmed this)

Reply:
ah, perfect!